File: Sorting/JavaScript/cyclicSortBasic.js
// sort the given array using cyclic sort
// one condition that we need to remember for cyclic sort is that the given an array is within the range 1-n

const arr = [2, 3, 5, 1, 4];
console.log("unsorted array: ", arr)


function swap (arr, firstIndex, secondIndex) {
  const temp = arr[firstIndex];
  arr[firstIndex] = arr[secondIndex];
  arr[secondIndex] = temp;
}

function sort (arr) {
  let i = 0;
  while (i < arr.length) {
    let correctIndex = arr[i] - 1;
    if (arr[correctIndex] !== arr[i]) {
      swap(arr, correctIndex, i)
    } else {
      i++;
    }
  } 
}

sort(arr)
console.log("sorted array: ", arr)

File: Sorting/src/com/vinayak/CyclicSortBasic.java
package com.vinayak;

import java.util.Arrays;

public class CyclicSortBasic {
    public static void main(String[] args) {
        int[] arr = {5, 3 ,1, 2, 4};
        System.out.println("Unsorted array: " + Arrays.toString(arr));
        sort(arr);
        System.out.println("Sorted array: " + Arrays.toString(arr));
    }

    static void sort(int[] arr) {
        int i = 0;
        while (i < arr.length) {
            int correctIndex = arr[i] - 1;
            if (arr[i] != arr[correctIndex]) {
                swap(arr, correctIndex, i);
            } else {
                i++;
            }
        }
    }

    static void swap(int[] arr, int firstIndex, int secondIndex) {
        int temp = arr[firstIndex];
        arr[firstIndex] = arr[secondIndex];
        arr[secondIndex] = temp;
    }
}
